File: .agents/skills/find-distributor/scripts/extract.ts
import fs from 'node:fs';
import * as xlsx from 'xlsx';

export function extractBrands(filePath: string): string[] {
  if (!fs.existsSync(filePath)) {
    console.error(`Ошибка: Файл ${filePath} не найден.`);
    process.exit(1);
  }

  try {
    const workbook = xlsx.readFile(filePath);
    const sheetName = workbook.SheetNames[0];
    if (!sheetName) throw new Error("Excel файл не содержит листов.");
    
    const sheet = workbook.Sheets[sheetName];
    
    // В xlsx.utils.sheet_to_json можно передать тип, но проще использовать Record<string, unknown>
    const data = xlsx.utils.sheet_to_json<Record<string, unknown>>(sheet, { defval: '' });
    
    if (data.length === 0) {
      console.error("Ошибка: Файл пуст или имеет неверный формат.");
      process.exit(1);
    }
    
    const keys = Object.keys(data[0] || {});
    const getRealKey = (searchKey: string) => keys.find(k => k.toLowerCase().trim() === searchKey);
    
    const brandCol = getRealKey('brand');
    const manufacturerCol = getRealKey('manufacturer');

    if (!brandCol && !manufacturerCol) {
      console.warn("Предупреждение: Колонки 'brand' и 'manufacturer' не найдены по имени.");
    }
    
    const brands = new Set<string>();
    
    data.forEach(row => {
      if (brandCol && row[brandCol]) {
        brands.add(String(row[brandCol]).trim());
      }
      if (manufacturerCol && row[manufacturerCol]) {
        brands.add(String(row[manufacturerCol]).trim());
      }
    });
    
    const ignoreList = ['nan', 'none', 'unknown', '-', ''];
    const cleanedBrands = Array.from(brands)
      .filter(b => b && !ignoreList.includes(b.toLowerCase()))
      .sort();
      
    return cleanedBrands;
    
  } catch (e: unknown) {
    const errorMsg = e instanceof Error ? e.message : String(e);
    console.error(`Ошибка при обработке файла: ${errorMsg}`);
    process.exit(1);
  }
}

// Проверяем, запущен ли скрипт напрямую
const isMainModule = process.argv[1] && fs.realpathSync(__filename) === fs.realpathSync(process.argv[1]);

if (isMainModule) {
  const filePath = process.argv[2];
  if (!filePath) {
    console.error("Пожалуйста, укажите путь к Excel файлу. Например: npx tsx extract.ts keepa/data.xlsx");
    process.exit(1);
  }
  
  console.log(`Чтение файла ${filePath}...`);
  const brands = extractBrands(filePath);
  
  console.log(`\n✅ Найдено уникальных брендов/производителей: ${brands.length}`);
  console.log("-".repeat(40));
  brands.forEach(b => console.log(b));
  console.log("-".repeat(40));
}

File: .agents/skills/find-distributor/scripts/extract.py
import pandas as pd
import argparse
import sys
import os

def extract_brands(file_path):
    if not os.path.exists(file_path):
        print(f"Ошибка: Файл {file_path} не найден.")
        sys.exit(1)
        
    print(f"Чтение файла {file_path}...")
    try:
        # Читаем Excel файл
        df = pd.read_excel(file_path, engine='openpyxl')
        
        # Ищем колонки 'manufacturer' и 'brand' (учитываем разный регистр)
        cols_lower = {str(c).lower().strip(): c for c in df.columns}
        
        brand_col = cols_lower.get('brand')
        manufacturer_col = cols_lower.get('manufacturer')
        
        # Если по имени не нашли, пробуем по индексам колонок (DR = 121, DS = 122, если 0-индексация)
        if not brand_col and not manufacturer_col:
            print("Предупреждение: Колонки 'brand' и 'manufacturer' не найдены по имени. Пробуем найти по индексам колонок DR и DS.")
            if len(df.columns) > 122:
                brand_col = df.columns[122] # DS
                manufacturer_col = df.columns[121] # DR
            else:
                print("Ошибка: В файле недостаточно колонок для поиска по индексам DR и DS.")
                sys.exit(1)
            
        brands = set()
        
        if brand_col:
            brands.update(df[brand_col].dropna().astype(str).unique())
            
        if manufacturer_col:
            brands.update(df[manufacturer_col].dropna().astype(str).unique())
            
        # Очистка от пустых значений и мусора
        ignore_list = ['nan', 'none', 'unknown', '-', '']
        cleaned_brands = sorted({
            b.strip() for b in brands 
            if b.strip() and b.strip().lower() not in ignore_list
        })
        
        print(f"\n✅ Найдено уникальных брендов/производителей: {len(cleaned_brands)}")
        print("-" * 40)
        for b in cleaned_brands:
            print(b)
        print("-" * 40)
            
    except Exception as e:
        print(f"Ошибка при обработке файла: {e}")
        sys.exit(1)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Извлечение уникальных брендов из выгрузки Keepa.")
    parser.add_argument("file_path", help="Путь к файлу Excel")
    args = parser.parse_args()
    
    extract_brands(args.file_path)
